### PR TITLE
massdriver-application-gcp-cloud-function

### PR DESCRIPTION
**Pull request by WillBeebe**
**IGNORE** all the files in the `files/*` folder. Those are needed to provision the function the first time for each runtime. This adds a lot of lines to this PR, it isn't as scary as it looks. :)

resolves: https://github.com/massdriver-cloud/massdriver/issues/1459

- deleting the `gcp-cloud-function` module. Cloud functions as app templates make more sense and we don't need that module anymore.

### CODE REVIEW

File: massdriver-application-gcp-cloud-function/artifacts.tf [hunk 1]
```diff
@@ -0,0 +1,28 @@
+locals {
+  data_infrastructure = {
```
Comment (WillBeebe) [inline massdriver-application-gcp-cloud-function/artifacts.tf:2]: work in progress with what we had from functions as infra bundles. Ready to match it with app artifacts when that is ready.

File: massdriver-application-gcp-cloud-function/bucket.tf [hunk 2]
```diff
@@ -0,0 +1,18 @@
+resource "google_storage_bucket" "main" {
```
Comment (WillBeebe) [inline massdriver-application-gcp-cloud-function/bucket.tf:1]: Cloud Functions needs source code in order to provision. For the first provision and subsequent ones with user code, this bucket will be the storage destination. Kinda like a container repository for zip files.

File: massdriver-application-gcp-cloud-function/bucket.tf [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  name                        = var.md_metadata.name_prefix
+  labels                      = var.md_metadata.default_tags
+  location                    = var.location
+  uniform_bucket_level_access = true
+}
+
+data "archive_file" "placeholder" {
+  type        = "zip"
+  source_dir  = "${path.module}/files/${var.cloud_function_configuration.runtime}"
```
Comment (WillBeebe) [inline massdriver-application-gcp-cloud-function/bucket.tf:10]: the runtime determines which folder we zip

File: massdriver-application-gcp-cloud-function/files/python310/main.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
+import logging
+import os
+
+import redis
+from pyramid.config import Configurator
+from pyramid.response import Response
+
+log = logging.getLogger(__name__)
```
Comment (WillBeebe) [inline massdriver-application-gcp-cloud-function/files/python310/main.py:8]: TODO: replace this with hello world. Move this to testing / examples.

File: massdriver-application-gcp-cloud-run/main.tf [hunk 5]
```diff
@@ -23,6 +23,9 @@ resource "google_cloud_run_service" "main" {
       service_account_name = module.application.id
       containers {
         image = var.container_image
+        ports {
+          container_port = var.container_port
```
Comment (WillBeebe) [inline massdriver-application-gcp-cloud-run/main.tf:27]: Cloud Run encourages the user to listen to the `PORT` env-var. This is to support container images that do not (like ones made from cloud functions buildpacks).

File: massdriver-application-gcp-cloud-function/main.tf [hunk 6]
```diff
@@ -0,0 +1,36 @@
+module "application" {
+  source  = "github.com/massdriver-cloud/terraform-modules//massdriver-application"
+  name    = var.md_metadata.name_prefix
+  service = "function"
+}
+
+resource "google_cloudfunctions_function" "main" {
```
Comment (WillBeebe) [inline massdriver-application-gcp-cloud-function/main.tf:7]: main resource to review

**Review by xpositivityx — APPROVED** [review]

**Review by mclacore — COMMENTED** [review]
Just the artifact file, everything else looks great

File: massdriver-application-gcp-cloud-function/artifacts.tf [hunk 7]
```diff
@@ -0,0 +1,28 @@
+locals {
```
[the remark this answers is not in the document]
Comment (WillBeebe) [reply massdriver-application-gcp-cloud-function/artifacts.tf:1]: That's a great point! I'll yank it from this PR 👍

**Review by mclacore — APPROVED** [review]
LGTM

**Pull request merged**